### PR TITLE
Userobject to check element quality at execute_on = initial (#10377)

### LINKED ISSUE #10377: Add element quality checker
## Rationale

Element quality such as distortion or stretch should be checked before simulation run.
## Description

ElementQualityChecker is a user object based on libMesh quality metrics to check the quality of each element at execute_on =  initial. Two type of failure modes are available: WARNING or Error. Users can specify the failure bounds for each available quality metric, otherwise default values from libMesh are used.
## Impact

New capability. No impact on existing codes.

### PR DESCRIPTION
**Pull request by hchen139**
closes #10377

### CODE REVIEW

**Comment by hchen139** [comment]
This is the element quality checker @bwspenc

File: framework/include/userobject/ElementQualityChecker.h
Comment (permcody) [inline framework/include/userobject/ElementQualityChecker.h]: Wrong header for the framework objects.

File: framework/include/userobject/ElementQualityChecker.h
Comment (permcody) [inline framework/include/userobject/ElementQualityChecker.h]: dof_id_type

File: framework/include/userobject/ElementQualityChecker.h
Comment (permcody) [inline framework/include/userobject/ElementQualityChecker.h]: Bad idea. Using Reals for keys causes lots of problems. The problem is the comparison operator when inserting values that are close to one and other.

File: framework/include/userobject/ElementQualityChecker.h
Comment (permcody) [inline framework/include/userobject/ElementQualityChecker.h]: Probably not necessary, I'll explain in the context below.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: Why not just use default values in the param object?

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: This check isn't good enough. If you only supply a `_lower_bound` and it's positive, you won't hit this for instance.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: You only have one value for bypassed but you could have mixed types where some are applicable and some are not. You don't know when visiting which order you will hit these elements in. It's possible to hit elements that you do want to check after elements that are bypassed. This logic is faulty in that case.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: So there's no correlation between element IDs and these values? If you just want a sorted list of elements you should probably just insert them into a vector (or list) something where the key is not used to build the data structure (since it is a real). You can always just sort once in finalize. Not only will it perform better, but it'll also just work where Reals as keys can fail.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: So this is up to you if you really want to control element metrics separately from everything else in MOOSE, but we have another mechanism you can consider. `mooseWarning` already gets promoted to `mooseError` when you pass `--error` on the command line. We use that in our test suite and in many other places to trigger errors when warnings are printed. We created that system to avoid this very kind of logic here. It's all built-in so you can remove dozens of lines of code.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: Use `_console` instead. It's buffered and can be turned off.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: We have a `stringify` utility for printing containers out. You might want to try it.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: Both of these parameters should be passed in by const reference.

**Review by permcody — CHANGES_REQUESTED** [review]
@jwpeterson - Under which conditions do quality metrics get printed from libMesh now? Is adding a new UO overkill?

**Comment by jwpeterson** [comment]
> @jwpeterson - Under which conditions do quality metrics get printed from libMesh now? Is adding a new UO overkill?

Nothing prints them right now--that is the point of adding this UO I believe.

File: framework/src/userobject/ElementQualityChecker.C
Comment (bwspenc) [reply framework/src/userobject/ElementQualityChecker.C]: This was my idea. We are planning to have an Action that sets up a few of these quality checks by default for an application code. The intent here was that we would have the Action set up two copies of some of these objects: one with a warning threshold and one with an error threshold (to completely refuse to run with more eggregious violations of the quality checks). Since we're trying to stop users from running with bad meshes, I don't think that relying them to use '--error' in this case makes sense.

**Comment by bwspenc** [comment]
@jwpeterson is correct.  The purpose of this is to impose restrictions on the mesh quality in an application code.  We will add an Action that sets up a set of these objects automatically, so users will get this check whether they want it or not.

File: framework/src/userobject/ElementQualityChecker.C
Comment (bwspenc) [reply framework/src/userobject/ElementQualityChecker.C]: @permcody You're correct, that would be better. Handling the defaults is a little tricker than that here, though. Every one of these metrics has different default thresholds per element type in libmesh. I think the right way to handle this is to defer to those defaults if 'upper_bound' or 'lower_bound' is not specified by the user, but use the user-supplied values if they exist.

File: framework/include/userobject/ElementQualityChecker.h
Comment (bwspenc) [reply framework/include/userobject/ElementQualityChecker.h]: I agree.  We just talked, and it doesn't need to be done this way.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [reply framework/src/userobject/ElementQualityChecker.C]: I'm still confused as to what _bypassed is supposed to do. You are setting it to 1 here, but you are summing on it in threadJoin and finalize. That means it's dependent on the number of processors you are running on. That can't be right.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [inline framework/src/userobject/ElementQualityChecker.C]: Is it your intent that these come out after the error message? They won't if you put them here. They will for mooseWarning since that returns, but not for mooseError().

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [reply framework/src/userobject/ElementQualityChecker.C]: You need to initialize all of your Reals in the constructor too. I know this may seem redundant since you reset them in initialize, but this is just good defensive coding. If your UO gets turned off and initialize never gets called but for some reason another public method does to retrieve those values, they'd be in an invalid state. Just get in the habit of initializing all POD types.

File: test/tests/userobjects/element_quality_check/tests
Comment (permcody) [inline test/tests/userobjects/element_quality_check/tests]: Please indent this file like you do the input files.

File: framework/src/userobject/ElementQualityChecker.C
Comment (permcody) [reply framework/src/userobject/ElementQualityChecker.C]: Oh, Since you are intentionally not using param defaults. It might be worth putting in a note as to why just so it's clear to another developer.

**Review by permcody — CHANGES_REQUESTED** [review]
This looks better. I have just a few more comments before this gets merged in.

File: framework/src/userobject/ElementQualityChecker.C
Comment (hchen139) [reply framework/src/userobject/ElementQualityChecker.C]: @permcody The logic behind the summation of value of _bypassed is to check whether there is element whose quality check was bypassed across all processors. If so, the value of _bypassed will be nonzero. Otherwise its value is zero. Depends on the value of this summation, we decide whether there is need to print out the bypassed element type or not. At least this is my intension, if the current code doesn't work as what I intent.

Comment (permcody) [reply framework/src/userobject/ElementQualityChecker.C]: So what you are saying is that _bypass is actually only used as a Boolean, not a numeric value. This means it should actually be declared and used as a Boolean. We do this kind of thing in the framework all over the place.

Aggregation:
You can use `|=` or `&=` to aggregate Booleans in threadJoin routines.
You can use Communicator::max/min to aggregate Boolean values in `finalize`.

I think this approach is cleaner than using ints. Your intention is clear and the value is stable in parallel, etc.

**Review by permcody — COMMENTED** [review]
I think this is all OK. Is there any interest in making an Auxkernel that displays these metrics as fields too?

**Review by permcody — APPROVED** [review]

**Comment by brianmoose** [comment]
It looks like this PR broke our PBS test: https://www.moosebuild.org/job/140736/ , although all the failed tests should fail with our normal tests.

The reason seems to be the `failure_type` parameter, which has an option of being `ERROR`.
The yaml dump test (and others) fail because the TestHarness checks to see if `ERROR` is in the output and assumes if it is then an error occurred. 
The reason the normal tests pass is because the TestHarness automatically trims large output and the `failure_type` parameter was part of the output that got cut.
When running in PBS mode, no output is trimmed, and the RunApp tester sees `ERROR` in the output, which causes the failure.

**Comment by permcody** [comment]
Holy crap... Well that's not convenient!

**Comment by bwspenc** [comment]
So no part of the parameter can contain the string 'ERROR'?

**Comment by permcody** [comment]
It sounds like it's time to rethink that it means to have an error. It was probably my crumby idea to look for a string like that and assume failure. The return codes have been reliable for a long time. I'd be willing to back off from the string matching.

**Comment by brianmoose** [comment]
FYI, here are the strings that it checks for anywhere in the output.
'ERROR', 'command not found', 'erminate called after throwing an instance of'

**Comment by permcody** [comment]
I bet that every one of those is a non-zero error code, what do you think?

**Comment by brianmoose** [comment]
Yeah, most of the time those should return a bad exit code.
Exit codes are far more reliable than looking for strings in the output.
The other (quick and dirty) option is to just zero out the "errors" parameter in the spec file for the various dump tests that are failing.

**Comment by brianmoose** [comment]
Looks like `outputs.format.no_double_executioner_output`
sets the "errors" parameter to 'zzzzz zzzzz' for this exact same reason. ERROR could occur in the output.

**Comment by permcody** [comment]
Hmm... That's not a bad workaround.

**Pull request merged**
